### PR TITLE
Avoid vanity alignment in base ctor call

### PR DESCRIPTION
**Pull request by su8898**
Fixes https://github.com/fsprojects/fantomas/issues/1442

### CODE REVIEW

File: src/Fantomas/CodePrinter.fs [hunk 1]
```diff
@@ -2526,6 +2545,8 @@ and genExpr astContext synExpr ctx =
 
     expr ctx
 
+and genExprInherit astContext synExpr ctx node = genExpr astContext synExpr ctx
```
Comment (nojaf) [inline src/Fantomas/CodePrinter.fs:2548]: Please don't introduce unused code.
A PR here should contain the bare minimum change that solve an issue.

File: src/Fantomas.Tests/TypeDeclarationTests.fs [hunk 2]
```diff
@@ -1708,7 +1708,11 @@ type VersionMismatchDuringDeserializationException
         message : string,
         innerException : System.Exception
     ) =
-    inherit System.Exception(message, innerException)
+    inherit System.Exception
```
Comment (nojaf) [inline src/Fantomas.Tests/TypeDeclarationTests.fs:1711]: This test should not be updated, as it doesn't cross the max_line_length here.

File: src/Fantomas/CodePrinter.fs [hunk 3]
```diff
@@ -29,15 +29,18 @@ type ASTContext =
       /// First type param might need extra spaces to avoid parsing errors on `<^`, `<'`, etc.
       IsFirstTypeParam: bool
       /// Inside a SynPat of MatchClause
-      IsInsideMatchClausePattern: bool }
+      IsInsideMatchClausePattern: bool
+      /// Is a call to base class constructor
+      IsBaseCtorCall: bool }
```
Comment (nojaf) [inline src/Fantomas/CodePrinter.fs:34]: Please do not extend ASTContext, that is last resort solution that should always be avoided.
This has lead to many bugs and can be avoided here.

File: src/Fantomas/CodePrinter.fs [hunk 4]
```diff
@@ -4555,7 +4576,10 @@ and genMemberDefn astContext node =
         !- "inherit "
         +> genType astContext false t
         +> addSpaceBeforeClassConstructor e
-        +> genExpr astContext e
+        +> genExpr
```
Comment (nojaf) [inline src/Fantomas/CodePrinter.fs:4579]: The code fix should be in this area.
I propose you pattern match on the expression and create a separate case if it `SynExpr.Paren(..)` that is too long.
Use the `expressionFitsOnRestOfLine` to determine if the short or long path should be followed.

**Review by nojaf — CHANGES_REQUESTED** [review]
Hello there, thank you for showing interest in this project.
I'm afraid the fix you had in mind isn't really going to fly.
I should add a note to the [contributing guidelines](https://github.com/fsprojects/fantomas/blob/master/CONTRIBUTING.md) that extending ASTContext should be avoided.
Let me know if you have any more questions.

**Comment by nojaf** [comment]
Also, please rebase on the latest master.

File: src/Fantomas/CodePrinter.fs (hunk 4 above)
Comment (su8898) [reply src/Fantomas/CodePrinter.fs:4579]: Hey, many thanks for your notes and suggestions. I will go through them.

**Comment by su8898** [comment]
@nojaf Hi, I've updated the PR. Please have a look again. Thank you.

File: src/Fantomas/CodePrinter.fs [hunk 5]
```diff
@@ -913,12 +913,17 @@ and genAnonRecordFieldName astContext (AnonRecordFieldName (s, e)) =
 
     !-s +> sepEq +> expr
 
-and genTuple astContext es =
+and genTuple astContext es isBaseCtorCall =
```
Comment (nojaf) [inline src/Fantomas/CodePrinter.fs:916]: Don't change `genTuple` here, your edge case is in `MDImplicitInherit`. It should be tackle in full over there.

File: src/Fantomas.Tests/ClassTests.fs [hunk 6]
```diff
@@ -896,3 +896,38 @@ type MaybeBuilder() =
         ) : 'U option =
         Option.bind binder value
 """
+
+[<Test>]
+let ``avoid vanity alignment when calling base constructor`` () =
```
Comment (nojaf) [inline src/Fantomas.Tests/ClassTests.fs:901]: Please add the issue number that the end of the test. See [Contribution guidelines](https://github.com/fsprojects/fantomas/blob/master/CONTRIBUTING.md#guidelines).

File: src/Fantomas/CodePrinter.fs [hunk 7]
```diff
@@ -4552,10 +4557,45 @@ and genMemberDefn astContext node =
     | MDOpen s -> !-(sprintf "open %s" s)
     // What is the role of so
     | MDImplicitInherit (t, e, _) ->
+        let genBasecall =
+            let shortExpr = genExpr astContext e
+
+            let longExpr =
+                match e with
```
Comment (nojaf) [inline src/Fantomas/CodePrinter.fs:4564]: You are doing a bit much here, I would go for a more focused pattern match to make your test pass.
There are no tests for the cases where you would have a `LetOrUses` or `Sequential`.
This is fine, but again don't write more code than necessary.

I'd go with something like:
```fsharp
match e with 
| Paren (lpr, (Tuple(es, tr)), rpr, pr) ->
                        sepNln
                        +> indent
                        +> sepOpenTFor lpr
                        +> sepNln
                        +> (col (sepComma +> sepNln) es (genExpr astContext) |> genTriviaFor SynExpr_Tuple tr)
                        +> unindent
                        +> sepNln
                        +> unindent
                        +> sepCloseTFor rpr pr
                        |> genTriviaFor SynExpr_Paren pr
| _ -> genExpr astContext e
```

This is sudo code, I don't know if this snippet will work but you get the idea.
Please also notice that I've add function calls to `genTriviaFor`, this is necessary if the parenthesis or tuple were to have an trivia (like code comments).

**Review by nojaf — CHANGES_REQUESTED** [review]
You are getting there, keep it up.

**Comment by su8898** [comment]
> You are getting there, keep it up.

Many thanks for the comments and suggestions. I will go through them and update.

**Comment by su8898** [comment]
@nojaf Hi, I've updated the PR after incorporating your suggestions. Please have a look again. Thank you.

**Review by nojaf — APPROVED** [review]
Many thanks for these changes!
I've tweaked the unit test just so that is conform with all the others. (Couldn't help myself).|
A fine first contribution!

**Comment by su8898** [comment]
> Many thanks for these changes!
> I've tweaked the unit test just so that is conform with all the others. (Couldn't help myself).|
> A fine first contribution!

Thanks so much 😄
